internal/raft/storage/store.go

```diff
@@ -0,0 +1,412 @@
+	return index + 1, err
+}
+
+func (es *EntryStore) Snapshot() (pb.Snapshot, error) {
+	es.Lock()
+	defer es.Unlock()
+	return es.snapshot, nil
+}
+
+func (es *EntryStore) ApplySnapshot(snap pb.Snapshot) error {
+	es.Lock()
+	defer es.Unlock()
+
+	//handle check for old snapshot being applied
+	msIndex := es.snapshot.Metadata.Index
+	snapIndex := snap.Metadata.Index
+	if msIndex >= snapIndex {
+		return raft.ErrSnapOutOfDate
+	}
+	es.snapshot = snap
+	entries := []pb.Entry{{Term: snap.Metadata.Term, Index: snap.Metadata.Index}}
+	return es.replaceAllEntries(entries)
+}
+
+func (es *EntryStore) CreateSnapshot(i uint64, cs *pb.ConfState, data []byte) (pb.Snapshot, error) {
+	es.Lock()
+	defer es.Unlock()
+	if i <= es.snapshot.Metadata.Index {
+		return pb.Snapshot{}, raft.ErrSnapOutOfDate
+	}
+
+	lastIdx, err := es.fetchIndexLimit(true)
+	if err != nil {
+		return pb.Snapshot{}, err
+	}
+	if i > lastIdx {
+		log.Panicf("snapshot %d is out of bound lastindex(%d)", i, lastIdx)
+	}
+
+	term, err := es.fetchTermForIndex(i)
+	if err != nil {
+		return pb.Snapshot{}, err
+	}
+
+	es.snapshot.Metadata.Index = i
+	es.snapshot.Metadata.Term = term
+	if cs != nil {
+		es.snapshot.Metadata.ConfState = *cs
+	}
+	es.snapshot.Data = data
+	return es.snapshot, nil
+}
+
+func (es *EntryStore) Compact(compactIndex uint64) error {
+	es.Lock()
+	defer es.Unlock()
+	beginIndex, err := es.fetchIndexLimit(false)
+	if err != nil {
+		return err
+	}
+
+	if compactIndex <= beginIndex {
+		return raft.ErrCompacted
+	}
+
+	endIndex, err := es.fetchIndexLimit(true)
+	if err != nil {
+		return err
+	}
+
+	if compactIndex > endIndex {
+		log.Panicf("compact %d is out of bound lastindex(%d)", compactIndex, endIndex)
+	}
+	return es.removeEntriesFrom(compactIndex, true)
+}
+
+func (es *EntryStore) Append(entries []pb.Entry) error {
+	if len(entries) == 0 {
+		return nil
+	}
+
+	es.Lock()
+	defer es.Unlock()
+
+	last := entries[0].Index + uint64(len(entries)) - 1
+	firstStoredIndex, err := es.fetchIndexLimit(false)
+	if err != nil || last <= firstStoredIndex {
+		return err
+	}
+
+	lastStoredIndex, err := es.fetchIndexLimit(true)
+	if err != nil {
+		return err
+	}
+	if entries[0].Index > (lastStoredIndex + 1) {
+		log.Panicf("missing log entry [last: %d, append at: %d]",
+			lastStoredIndex, entries[0].Index)
+	}
+
+	/*
+		case 1: truncate compacted entries
+		input: 		2-3-4-5
+		exist:		  3-4-5
+		new_input:	4-5
+	 */
+	if firstStoredIndex >= entries[0].Index {
+		entries = entries[1+firstStoredIndex-entries[0].Index:]
+	}
+
+	/*
+		case 2: truncate existing entries
+		input:		  4'-5'-6'-7'-8'
+		exist:		3-4-5
+		new_exist:	3-4'-5'-6'-7'-8'
+	 */
+	if err := es.removeEntriesFrom(entries[0].Index, false); err != nil {
+		return err
+	}
+
+	/*
+		case 3: simple append
+		input: 		      4'-5'-6'
+		exist:		1-2-3
+		new_exist:	1-2-3-4'-5'-6'
+	 */
+	return es.appendEntries(entries)
+}
+
+func (es *EntryStore) Close() error {
+	es.Lock()
+	defer es.Unlock()
```

Comment:
do we need a lock in `close` ?

Reply:
Yes, in case an Append or Compact is already in progress, this lock prevents any possible corruption upon closing.